## let's study the csv files 

In [1]:
import pandas as pd

In [2]:
df_1 = pd.read_csv(r"C:\Users\peter\OneDrive\Desktop\final_exam\interviews\incidenti_stradali_ultimi_5_anni.csv")
df_1.shape

(90144, 16)

In [3]:
df_1.info()

<class 'pandas.DataFrame'>
RangeIndex: 90144 entries, 0 to 90143
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   DATAFLOW          90144 non-null  str    
 1   FREQ              90144 non-null  str    
 2   REF_AREA          90144 non-null  int64  
 3   DATA_TYPE         90144 non-null  str    
 4   RESULT            90144 non-null  str    
 5   TIME_PERIOD       90144 non-null  int64  
 6   OBS_VALUE         90144 non-null  int64  
 7   OBS_STATUS        0 non-null      float64
 8   NOTE_DS           0 non-null      float64
 9   NOTE_REF_AREA     0 non-null      float64
 10  NOTE_DATA_TYPE    0 non-null      float64
 11  NOTE_RESULT       0 non-null      float64
 12  NOTE_TIME_PERIOD  0 non-null      float64
 13  BASE_PER          0 non-null      float64
 14  UNIT_MEAS         0 non-null      float64
 15  UNIT_MULT         0 non-null      float64
dtypes: float64(9), int64(3), str(4)
memory usage: 11.0 

In [4]:
df_1.sample(10)

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,NOTE_DATA_TYPE,NOTE_RESULT,NOTE_TIME_PERIOD,BASE_PER,UNIT_MEAS,UNIT_MULT
47019,IT1:41_983(1.0),A,43053,KILLINJ,M,2023,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
41329,IT1:41_983(1.0),A,30190,KILLINJ,M,2024,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
27520,IT1:41_983(1.0),A,18168,KILLINJ,F,2022,6,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
64440,IT1:41_983(1.0),A,68038,KILLINJ,F,2021,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
80946,IT1:41_983(1.0),A,92050,KILLINJ,F,2024,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19120,IT1:41_983(1.0),A,15032,KILLINJ,M,2024,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
11580,IT1:41_983(1.0),A,7024,KILLINJ,M,2021,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
70132,IT1:41_983(1.0),A,76064,KILLINJ,F,2022,12,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
16916,IT1:41_983(1.0),A,13075,ROADACC,9,2024,406,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
974,IT1:41_983(1.0),A,1088,KILLINJ,F,2023,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## after reviewing df_1, i begin cleaning it.

In [5]:
df_1.columns

Index(['DATAFLOW', 'FREQ', 'REF_AREA', 'DATA_TYPE', 'RESULT', 'TIME_PERIOD',
       'OBS_VALUE', 'OBS_STATUS', 'NOTE_DS', 'NOTE_REF_AREA', 'NOTE_DATA_TYPE',
       'NOTE_RESULT', 'NOTE_TIME_PERIOD', 'BASE_PER', 'UNIT_MEAS',
       'UNIT_MULT'],
      dtype='str')

In [6]:
cars_accidents = df_1.copy()
cols_1 = ['DATAFLOW',
          'FREQ',
          'OBS_STATUS',
          'NOTE_DS',
          'NOTE_REF_AREA',
          'NOTE_DATA_TYPE',
          'NOTE_RESULT',
          'NOTE_TIME_PERIOD',
          'BASE_PER',
          'UNIT_MEAS',
          'UNIT_MULT']
cars_accidents.drop(columns=cols_1, inplace=True)
cars_accidents.shape

(90144, 5)

In [7]:
cars_accidents.info()

<class 'pandas.DataFrame'>
RangeIndex: 90144 entries, 0 to 90143
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   REF_AREA     90144 non-null  int64
 1   DATA_TYPE    90144 non-null  str  
 2   RESULT       90144 non-null  str  
 3   TIME_PERIOD  90144 non-null  int64
 4   OBS_VALUE    90144 non-null  int64
dtypes: int64(3), str(2)
memory usage: 3.4 MB


In [8]:
cars_accidents.sample(10)

,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE
32347,22157,ROADACC,9,2022,1
57613,62038,KILLINJ,M,2024,0
43262,35046,KILLINJ,M,2022,0
11036,6168,ROADACC,9,2021,6
83228,96007,KILLINJ,M,2022,0
88816,110002,KILLINJ,F,2022,480
89570,111057,ROADACC,9,2024,2
27125,18133,KILLINJ,F,2023,2
58635,63049,ROADACC,9,2022,2337
9310,5121,KILLINJ,M,2024,0


In [9]:
cars_accidents = cars_accidents.rename(columns={'REF_AREA': 'Id_code','DATA_TYPE':'Type','RESULT':'Outcome','TIME_PERIOD':'Year','OBS_VALUE':'Obs_value'})
cars_accidents.head(10)

,Id_code,Type,Outcome,Year,Obs_value
0,1001,KILLINJ,F,2021,2
1,1001,KILLINJ,F,2022,5
2,1001,KILLINJ,F,2023,6
3,1001,KILLINJ,F,2024,5
4,1001,KILLINJ,M,2021,0
5,1001,KILLINJ,M,2022,0
6,1001,KILLINJ,M,2023,1
7,1001,KILLINJ,M,2024,0
8,1001,ROADACC,9,2021,2
9,1001,ROADACC,9,2022,3


In [10]:
cars_accidents['Outcome'] = cars_accidents['Outcome'].replace('9','')
cars_accidents['Metric'] = cars_accidents['Type'] + cars_accidents['Outcome']
cars_accidents.head(10)

,Id_code,Type,Outcome,Year,Obs_value,Metric
0,1001,KILLINJ,F,2021,2,KILLINJF
1,1001,KILLINJ,F,2022,5,KILLINJF
2,1001,KILLINJ,F,2023,6,KILLINJF
3,1001,KILLINJ,F,2024,5,KILLINJF
4,1001,KILLINJ,M,2021,0,KILLINJM
5,1001,KILLINJ,M,2022,0,KILLINJM
6,1001,KILLINJ,M,2023,1,KILLINJM
7,1001,KILLINJ,M,2024,0,KILLINJM
8,1001,ROADACC,,2021,2,ROADACC
9,1001,ROADACC,,2022,3,ROADACC


In [11]:
cars_accidents = cars_accidents.drop(columns=['Type','Outcome'])
cars_accidents['Metric'] = cars_accidents['Metric'].replace({'KILLINJF':'injured', 'KILLINJM' : 'deaths', 'ROADACC' : 'road_acc'})
cars_accidents = cars_accidents[['Id_code','Year','Metric','Obs_value']]
cars_accidents.head(10)

,Id_code,Year,Metric,Obs_value
0,1001,2021,injured,2
1,1001,2022,injured,5
2,1001,2023,injured,6
3,1001,2024,injured,5
4,1001,2021,deaths,0
5,1001,2022,deaths,0
6,1001,2023,deaths,1
7,1001,2024,deaths,0
8,1001,2021,road_acc,2
9,1001,2022,road_acc,3


In [12]:
cars_accidents_2021_2025 = cars_accidents.pivot_table(index=['Id_code','Year'],columns=['Metric'], values='Obs_value').reset_index()
cars_accidents_2021_2025.columns.name = None
c = ['deaths', 'injured', 'road_acc']
cars_accidents_2021_2025[c] = cars_accidents_2021_2025[c].astype('int64')
cars_accidents_2021_2025.sample(10)

,Id_code,Year,deaths,injured,road_acc
15635,43044,2021,0,82,60
4428,9038,2023,0,5,3
27999,97004,2022,0,17,11
17307,56021,2023,2,51,40
21830,69090,2021,0,10,5
29210,103075,2022,1,13,12
11316,23068,2024,0,8,6
22460,72017,2024,1,88,62
6171,14040,2024,0,5,3
8943,18105,2022,0,0,0


In [13]:
df_2 = pd.read_csv(r"C:\Users\peter\OneDrive\Desktop\final_exam\interviews\comuni_situas_istat.csv", sep=';')
df_2.shape


(7903, 17)

In [14]:
df_2.info()

<class 'pandas.DataFrame'>
RangeIndex: 7903 entries, 0 to 7902
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype
---  ------                          --------------  -----
 0   Codice Ripartizione geografica  7903 non-null   int64
 1   Codice Regione                  7903 non-null   int64
 2   Codice Provincia (Storico)      7903 non-null   int64
 3   Codice Provincia/Uts            7903 non-null   int64
 4   Codice Comune (alfanumerico)    7903 non-null   int64
 5   Codice Comune (numerico)        7903 non-null   int64
 6   Comune                          7902 non-null   str  
 7   Comune (dizione straniera)      124 non-null    str  
 8   Sigla automobilistica           7811 non-null   str  
 9   Capoluogo di Provincia/Uts      7903 non-null   int64
 10  Capoluogo di Regione            7903 non-null   int64
 11  Popolazione legale              7903 non-null   int64
 12  Anno Censimento                 7903 non-null   int64
 13  Superficie (Km

In [15]:
df_2.sample(10)

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente)
6999,5,20,90,90,90082,90082,Viddalba,NaN,SS,0,0,1694,2011,"50,4089",2020,1623,2020
334,1,1,2,2,2043,2043,Civiasco,NaN,VC,0,0,265,2011,"7,3878",2020,239,2020
676,1,1,4,4,4198,4198,Ruffia,NaN,CN,0,0,350,2011,"7,5091",2020,359,2020
5496,4,13,66,66,66063,66063,Ortona dei Marsi,NaN,AQ,0,0,592,2011,"57,1661",2020,437,2020
6038,4,16,75,75,75067,75067,Sanarica,NaN,LE,0,0,1503,2011,"13,0223",2020,1443,2020
5670,4,13,69,69,69036,69036,Fresagrandinaria,NaN,CH,0,0,1056,2011,"25,1486",2020,908,2020
4679,3,12,58,258,58041,58041,Gavignano,NaN,RM,0,0,1956,2011,"15,0431",2020,1920,2020
5231,4,15,64,64,64076,64076,Quadrelle,NaN,AV,0,0,1893,2011,"6,9311",2020,1833,2020
1990,1,3,16,16,16160,16160,Pedrengo,NaN,BG,0,0,5799,2011,"3,5999",2020,5905,2020
1127,1,7,8,8,8022,8022,Civezza,NaN,IM,0,0,640,2011,"3,8809",2020,619,2020


In [16]:
comuni_italiani = df_2.copy()

In [17]:
comuni_italiani['Superficie (Kmq)'] = comuni_italiani['Superficie (Kmq)'].str.replace('.', '').str.replace(',', '.').astype('float64')
comuni_italiani['Superficie (Kmq)'].dtype
comuni_italiani['Superficie (Kmq)'].sample(10)

3337     37.9981
7731      3.2630
2727     51.4599
2863     10.4144
4437    525.7677
5540     20.5246
2240     19.1322
5804     19.4484
453       6.6500
1780      5.4619
Name: Superficie (Kmq), dtype: float64

In [18]:
comuni_italiani[comuni_italiani['Comune'].isnull()]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente)
164,1,1,1,201,1168,1168,NaN,NaN,TO,0,0,7998,2011,24.6422,2020,7849,2020


## I discovered that the single municipality labeled "NaN" (1168) actually corresponds to a town in the province of Turin called None! xD

In [19]:
comuni_italiani['Comune'] = comuni_italiani['Comune'].fillna('None')


In [20]:
comuni_italiani[comuni_italiani['Comune'] == 'None']

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente)
164,1,1,1,201,1168,1168,None,NaN,TO,0,0,7998,2011,24.6422,2020,7849,2020


In [21]:
region = {
    1: 'Piemonte',
    2: "Valle d'Aosta",
    3: 'Lombardia',
    4: 'Trentino-Alto Adige',    
    5: 'Veneto',            
    6: 'Friuli-Venezia Giulia',
    7: 'Liguria',
    8: 'Emilia-Romagna',
    9: 'Toscana',
    10: 'Umbria',
    11: 'Marche',
    12: 'Lazio',
    13: 'Abruzzo',
    14: 'Molise',            
    15: 'Campania',
    16: 'Puglia',
    17: 'Basilicata',
    18: 'Calabria',
    19: 'Sicilia',
    20: 'Sardegna'}

comuni_italiani['Regione'] = comuni_italiani['Codice Regione'].map(region)
comuni_italiani.sample(10)

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),Regione
2693,2,4,21,21,21046,21046,Malles Venosta/Mals,Mals,BZ,0,0,5086,2011,247.4250,2020,5170,2020,Trentino-Alto Adige
5285,4,15,65,65,65010,65010,Atena Lucana,NaN,SA,0,0,2288,2011,26.0072,2020,2392,2020,Campania
2901,2,4,22,22,22224,22224,Volano,NaN,TN,0,0,3125,2011,10.7401,2020,3020,2020,Trentino-Alto Adige
1385,1,3,12,12,12046,12046,Castiglione Olona,NaN,VA,0,0,7836,2011,6.9003,2020,7635,2020,Lombardia
3350,2,5,28,28,28010,28010,Barbona,NaN,PD,0,0,733,2011,8.5865,2020,617,2020,Veneto
1511,1,3,13,13,13047,13047,Carlazzo,NaN,CO,0,0,2972,2011,12.7329,2020,3103,2020,Lombardia
3339,2,5,27,227,27043,27043,Vigonovo,NaN,VE,0,0,9892,2011,12.8657,2020,9936,2020,Veneto
6505,4,18,80,280,80075,80075,San Pietro di Caridà,NaN,RC,0,0,1265,2011,48.0777,2020,1030,2020,Calabria
4069,3,11,43,43,43011,43011,Cessapalombo,NaN,MC,0,0,546,2011,27.5803,2020,443,2020,Marche
6197,4,17,77,77,77028,77028,Tricarico,NaN,MT,0,0,5669,2011,178.1606,2020,4890,2020,Basilicata


In [22]:
rip_geo = {1: 'Nord-Ovest',
           2: 'Nord-Est',
           3: 'Centro',
           4: 'Sud',
           5: 'Isole'}
comuni_italiani['Ripartizione Geografica'] = comuni_italiani['Codice Ripartizione geografica'].map(rip_geo)
comuni_italiani.sample(10)

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),Regione,Ripartizione Geografica
7131,2,6,93,93,93033,93033,Pordenone,NaN,PN,1,0,50583,2011,38.2114,2020,51403,2020,Friuli-Venezia Giulia,Nord-Est
517,1,1,4,4,4038,4038,Canosio,NaN,CN,0,0,82,2011,48.4476,2020,79,2020,Piemonte,Nord-Ovest
3782,2,8,35,35,35039,35039,Sant'Ilario d'Enza,NaN,RE,0,0,10939,2011,20.2331,2020,11159,2020,Emilia-Romagna,Nord-Est
2533,1,3,19,19,19065,19065,Paderno Ponchielli,NaN,CR,0,0,1480,2011,23.9619,2020,1331,2020,Lombardia,Nord-Ovest
7772,3,11,109,109,109027,109027,Montottone,NaN,FM,0,0,1011,2011,16.3789,2020,895,2020,Marche,Centro
761,1,1,5,5,5034,5034,Celle Enomondo,NaN,AT,0,0,480,2011,5.5928,2020,462,2020,Piemonte,Nord-Ovest
3975,3,11,41,41,41019,41019,Gabicce Mare,NaN,PU,0,0,5845,2011,4.9349,2020,5597,2020,Marche,Centro
4699,3,12,58,258,58061,58061,Monteflavio,NaN,RM,0,0,1399,2011,16.8416,2020,1214,2020,Lazio,Centro
1576,1,3,13,13,13163,13163,Novedrate,NaN,CO,0,0,2873,2011,2.9210,2020,2860,2020,Lombardia,Nord-Ovest
2004,1,3,16,16,16174,16174,Predore,NaN,BG,0,0,1859,2011,10.9556,2020,1823,2020,Lombardia,Nord-Ovest


In [23]:
comuni_italiani['Capoluogo di Provincia/Uts'] = comuni_italiani['Capoluogo di Provincia/Uts'].replace({0: 'No', 1: 'Yes'})
comuni_italiani['Capoluogo di Regione'] = comuni_italiani['Capoluogo di Regione'].replace({0: 'No', 1: 'Yes'})
comuni_italiani.sample(10)

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),Regione,Ripartizione Geografica
5449,4,13,66,66,66016,66016,Campotosto,NaN,AQ,No,No,586,2011,51.7245,2020,464,2020,Abruzzo,Sud
7270,5,20,95,95,95068,95068,Ulà Tirso,NaN,OR,No,No,580,2011,18.8464,2020,489,2020,Sardegna,Isole
1963,1,3,16,16,16132,16132,Mapello,NaN,BG,No,No,6453,2011,8.6573,2020,6841,2020,Lombardia,Nord-Ovest
3044,2,5,24,24,24018,24018,Caldogno,NaN,VI,No,No,11221,2011,15.8845,2020,11328,2020,Veneto,Nord-Est
7782,3,11,109,109,109037,109037,Sant'Elpidio a Mare,NaN,FM,No,No,16968,2011,50.5172,2020,16773,2020,Marche,Centro
1250,1,7,10,210,10009,10009,Campomorone,NaN,GE,No,No,7306,2011,25.9059,2020,6488,2020,Liguria,Nord-Ovest
258,1,1,1,201,1263,1263,Sestriere,NaN,TO,No,No,838,2011,25.9158,2020,922,2020,Piemonte,Nord-Ovest
1401,1,3,12,12,12062,12062,Cuveglio,NaN,VA,No,No,3397,2011,7.5272,2020,3266,2020,Lombardia,Nord-Ovest
2652,2,4,21,21,21005,21005,Avelengo/Hafling,Hafling,BZ,No,No,739,2011,27.3952,2020,794,2020,Trentino-Alto Adige,Nord-Est
4448,3,10,54,54,54035,54035,Norcia,NaN,PG,No,No,4915,2011,275.5700,2020,4617,2020,Umbria,Centro


In [24]:
comuni_italiani.columns

Index(['Codice Ripartizione geografica', 'Codice Regione',
       'Codice Provincia (Storico)', 'Codice Provincia/Uts',
       'Codice Comune (alfanumerico)', 'Codice Comune (numerico)', 'Comune',
       'Comune (dizione straniera)', 'Sigla automobilistica',
       'Capoluogo di Provincia/Uts', 'Capoluogo di Regione',
       'Popolazione legale', 'Anno Censimento', 'Superficie (Kmq)',
       'Anno (Superficie)', 'Popolazione residente',
       'Anno (Popolazione residente)', 'Regione', 'Ripartizione Geografica'],
      dtype='str')

In [25]:
comuni_italiani_clean = comuni_italiani[['Codice Ripartizione geografica',
                                         'Codice Regione',
                                         'Codice Comune (alfanumerico)',
                                         'Codice Comune (numerico)',
                                         'Ripartizione Geografica',
                                         'Regione',
                                         'Sigla automobilistica',
                                         'Comune',
                                         'Capoluogo di Provincia/Uts',
                                         'Capoluogo di Regione',
                                         'Popolazione legale',
                                         'Superficie (Kmq)',
                                         ]]

comuni_italiani_clean.sample(10)

,Codice Ripartizione geografica,Codice Regione,Codice Comune (alfanumerico),Codice Comune (numerico),Ripartizione Geografica,Regione,Sigla automobilistica,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq)
3881,2,8,37051,37051,Nord-Est,Emilia-Romagna,BO,San Benedetto Val di Sambro,No,No,4393,66.4675
147,1,1,1150,1150,Nord-Ovest,Piemonte,TO,Mercenasco,No,No,1267,12.6383
5105,4,15,63040,63040,Sud,Campania,NaN,Liveri,No,No,1679,2.7109
1998,1,3,16168,16168,Nord-Ovest,Lombardia,BG,Ponte Nossa,No,No,1913,5.5898
260,1,1,1265,1265,Nord-Ovest,Piemonte,TO,Settimo Torinese,No,No,46873,31.4505
2709,2,4,21062,21062,Nord-Est,Trentino-Alto Adige,BZ,Parcines/Partschins,No,No,3536,55.3974
4672,3,12,58034,58034,Centro,Lazio,RM,Colleferro,No,No,21574,26.9893
5398,4,15,65123,65123,Sud,Campania,SA,San Mauro Cilento,No,No,985,15.2807
6949,5,20,90032,90032,Isole,Sardegna,SS,Ittireddu,No,No,580,23.6867
1139,1,7,8034,8034,Nord-Ovest,Liguria,IM,Mendatica,No,No,205,30.6912


In [26]:
cars_accidents_2021_2025.to_csv("cars_accidents_2021_2025.csv", index=False)
comuni_italiani_clean.to_csv("comuni_italiani_clean.csv", index=False)

# after studying the two CSV files and cleaning and saving them
# 

In [27]:
#after studying the two CSV files and cleaning and saving them, I will merge them
istat_incidenti_stradali_comuni= pd.merge(cars_accidents_2021_2025, 
                                           comuni_italiani_clean,
                                           left_on='Id_code',
                                           right_on='Codice Comune (numerico)',
                                           how='left')
istat_incidenti_stradali_comuni.sample(10)

,Id_code,Year,deaths,injured,road_acc,Codice Ripartizione geografica,Codice Regione,Codice Comune (alfanumerico),Codice Comune (numerico),Ripartizione Geografica,Regione,Sigla automobilistica,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq)
22168,70081,2023,0,0,0,4.0,14.0,70081.0,70081.0,Sud,Molise,CB,Trivento,No,No,4812.0,73.6983
2432,4178,2021,0,0,0,1.0,1.0,4178.0,4178.0,Nord-Ovest,Piemonte,CN,Prunetto,No,No,471.0,14.3615
22646,73019,2022,6,80,27,4.0,16.0,73019.0,73019.0,Sud,Puglia,TA,Mottola,No,No,16241.0,213.9603
10560,22061,2021,1,30,19,2.0,4.0,22061.0,22061.0,Nord-Est,Trentino-Alto Adige,TN,Civezzano,No,No,3904.0,15.6663
9922,20064,2023,0,0,0,1.0,3.0,20064.0,20064.0,Nord-Ovest,Lombardia,MN,Sustinente,No,No,2240.0,26.2721
4796,10065,2022,0,0,0,1.0,7.0,10065.0,10065.0,Nord-Ovest,Liguria,GE,Valbrevenna,No,No,812.0,34.6707
29062,103032,2023,0,0,0,1.0,1.0,103032.0,103032.0,Nord-Ovest,Piemonte,VB,Germagno,No,No,186.0,2.8976
18347,60007,2024,1,39,21,3.0,12.0,60007.0,60007.0,Centro,Lazio,FR,Aquino,No,No,5309.0,19.2368
9072,18141,2021,0,1,1,1.0,3.0,18141.0,18141.0,Nord-Ovest,Lombardia,PV,Sant'Alessio con Vialone,No,No,840.0,6.5572
6172,14041,2021,0,0,0,1.0,3.0,14041.0,14041.0,Nord-Ovest,Lombardia,SO,Mello,No,No,1001.0,11.4294


In [28]:
print(cars_accidents_2021_2025.shape)
print(comuni_italiani_clean.shape)
print(istat_incidenti_stradali_comuni.shape)

(30048, 5)
(7903, 12)
(30048, 17)


In [29]:
istat_incidenti_stradali_comuni.info()

<class 'pandas.DataFrame'>
RangeIndex: 30048 entries, 0 to 30047
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Id_code                         30048 non-null  int64  
 1   Year                            30048 non-null  int64  
 2   deaths                          30048 non-null  int64  
 3   injured                         30048 non-null  int64  
 4   road_acc                        30048 non-null  int64  
 5   Codice Ripartizione geografica  30030 non-null  float64
 6   Codice Regione                  30030 non-null  float64
 7   Codice Comune (alfanumerico)    30030 non-null  float64
 8   Codice Comune (numerico)        30030 non-null  float64
 9   Ripartizione Geografica         30030 non-null  str    
 10  Regione                         30030 non-null  str    
 11  Sigla automobilistica           29663 non-null  str    
 12  Comune                          30030 non-n

In [33]:
istat_incidenti_stradali_comuni[istat_incidenti_stradali_comuni['Comune'].isnull()]

,Id_code,Year,deaths,injured,road_acc,Codice Ripartizione geografica,Codice Regione,Codice Comune (alfanumerico),Codice Comune (numerico),Ripartizione Geografica,Regione,Sigla automobilistica,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq)
3105,5122,2023,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5458,12144,2023,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6019,13256,2024,0,4,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
11881,24128,2024,0,9,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
12115,25075,2024,0,32,20,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
13067,28108,2024,0,6,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24935,81025,2021,0,17,9,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24936,81025,2022,0,10,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24937,81025,2023,0,20,11,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
24938,81025,2024,0,17,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
istat_incidenti_stradali_comuni[istat_incidenti_stradali_comuni[']